packages/server/src/queries/complex/portfolio/allocation.ts

```diff
@@ -0,0 +1,184 @@
+
+  // Get top 5 assets by cap value
+  const sortedAccountCoinsResults =
+    totalAssets?.account_coins_result?.sort(
+      (a: AccountCoinsResult, b: AccountCoinsResult) =>
+        +b.cap_value - +a.cap_value
+    ) || [];
+
+  const top5AccountCoinsResults = sortedAccountCoinsResults.slice(0, 5);
+
+  const assets = top5AccountCoinsResults.map((asset: AccountCoinsResult) => {
+    const assetFromAssetLists = getAsset({
+      assetLists,
+      anyDenom: asset.coin.denom,
+    });
+
+    return {
+      key: assetFromAssetLists.coinDenom,
+      percentage: (+asset.cap_value / totalCap) * 100,
+      amount: +asset.cap_value,
+    };
+  });
+
+  const otherAssets = sortedAccountCoinsResults.slice(5);
+  const otherAmount = otherAssets.reduce(
+    (sum: number, asset: AccountCoinsResult) => sum + +asset.cap_value,
+    0
+  );
+  const otherPercentage = (otherAmount / totalCap) * 100;
+
+  const other = {
+    key: "Other",
+    percentage: otherPercentage,
+    amount: otherAmount,
+    color: "bg-osmoverse-500",
+  };
+
+  return [...assets, other];
+}
+
+async function getAvailable(categories: Categories, assetLists: AssetList[]) {
+  const userBalances = categories["user-balances"];
+  const totalCap = +userBalances.capitalization;
+
+  // Get top 5 assets by cap value
+  const sortedAccountCoinsResults =
+    userBalances?.account_coins_result?.sort(
+      (a: AccountCoinsResult, b: AccountCoinsResult) =>
+        +b.cap_value - +a.cap_value
+    ) || [];
+
+  const top5AccountCoinsResults = sortedAccountCoinsResults.slice(0, 5);
+
+  const assets = top5AccountCoinsResults.map((asset: AccountCoinsResult) => {
+    const assetFromAssetLists = getAsset({
+      assetLists,
+      anyDenom: asset.coin.denom,
+    });
+
+    return {
+      key: assetFromAssetLists.coinDenom,
+      percentage: (+asset.cap_value / totalCap) * 100,
+      amount: +asset.cap_value,
+    };
+  });
+
+  const otherAssets = sortedAccountCoinsResults.slice(5);
+  const otherAmount = otherAssets.reduce(
+    (sum: number, asset: AccountCoinsResult) => sum + +asset.cap_value,
+    0
+  );
+  const otherPercentage = (otherAmount / totalCap) * 100;
+
+  const other = {
+    key: "Other",
+    percentage: otherPercentage,
+    amount: otherAmount,
+  };
+
+  return [...assets, other];
+}
+
+export async function getAllocation({
+  address,
+  assetLists,
+}: {
+  address: string;
+  assetLists: AssetList[];
+}): Promise<GetAllocationResponse> {
+  return await cachified({
+    cache: allocationCache,
+    ttl: 1000 * 60 * 0.25, // 15 seconds since a user can transact quickly
+    key: `allocation-${address}`,
+    getFreshValue: async () => {
+      const data = await queryAllocation({
+        address,
+      });
+
+      const categories = data.categories;
+
+      const all = await getAll(categories);
+      const assets = await getAssets(categories, assetLists);
+      const available = await getAvailable(categories, assetLists);
```

Comment:
Nit: I don't think these need to be async functions. Could clean things up a bit.